### PR TITLE
fix: handle null ggv2dataclient

### PR DESCRIPTION
**Pull request by J-C-L**
**Issue #, if available:**

**Description of changes:**
Updating users of GreengrassV2DataClient getGreengrassV2DataClient() to handle a null response.

**Why is this change necessary:**
We've seen NPEs when the client is not retrieved.

**How was this change tested:**
- [X] Updated or added new unit tests.
- [ ] Updated or added new integration tests.
- [ ] Updated or added new end-to-end tests.
- [ ] If my code makes a remote network call, it was tested with a proxy.

**Any additional information or context required to review the change:**

**Documentation Checklist:**
 - [ ] Updated the README if applicable.

**Compatibility Checklist:**
- [X ] I confirm that the change is backwards compatible.
- [ ] Any modification or deletion of public interfaces does not impact other plugin components.
- [ ] For external library version updates, I have reviewed its change logs and Nucleus does not consume 
  any deprecated method or type.

Refer to [Compatibility Guidelines](/COMPATIBILITY.md) for more information.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: src/main/java/com/aws/greengrass/componentmanager/ComponentServiceHelper.java [hunk 1]
```diff
@@ -65,8 +72,24 @@ ResolvedComponentVersion resolveComponentVersion(String componentName, Semver lo
...
+        RetryUtils.RetryConfig clientExceptionRetryConfig =
+                RetryUtils.RetryConfig.builder().initialRetryInterval(Duration.ofSeconds(30))
+                        .maxRetryInterval(Duration.ofSeconds(30)).maxAttempt(CLIENT_RETRY_COUNT)
+                        .retryableExceptions(Arrays.asList(DeviceConfigurationException.class)).build();
+
+        try (GreengrassV2DataClient greengrasV2DataClient = RetryUtils.runWithRetry(clientExceptionRetryConfig, () -> {
+            GreengrassV2DataClient client = clientFactory.getGreengrassV2DataClient();
+            if (client == null) {
+                String errorMessage = clientFactory.getConfigValidationError().isPresent()
+                        ? clientFactory.getConfigValidationError().get() : "Could not get GreengrassV2DataClient";
```
Comment (rbattle) [inline src/main/java/com/aws/greengrass/componentmanager/ComponentServiceHelper.java:85]: nit: you could combine this as `clientFactory.getConfigValidationError().orElse("Could not get GreengrassV2DataClient")`

File: src/main/java/com/aws/greengrass/componentmanager/ComponentServiceHelper.java [hunk 2]
```diff
@@ -65,8 +72,24 @@ ResolvedComponentVersion resolveComponentVersion(String componentName, Semver lo
...
-        try {
-            result = clientFactory.getGreengrassV2DataClient().resolveComponentCandidates(request);
+
+        RetryUtils.RetryConfig clientExceptionRetryConfig =
+                RetryUtils.RetryConfig.builder().initialRetryInterval(Duration.ofSeconds(30))
+                        .maxRetryInterval(Duration.ofSeconds(30)).maxAttempt(CLIENT_RETRY_COUNT)
+                        .retryableExceptions(Arrays.asList(DeviceConfigurationException.class)).build();
+
+        try (GreengrassV2DataClient greengrasV2DataClient = RetryUtils.runWithRetry(clientExceptionRetryConfig, () -> {
+            GreengrassV2DataClient client = clientFactory.getGreengrassV2DataClient();
```
Comment (rbattle) [inline src/main/java/com/aws/greengrass/componentmanager/ComponentServiceHelper.java:82]: would it be better to keep and deprecate this API and add a new one that `throws DeviceConfigurationException`?

I'm not sure what a useful name for that API would be though

Comment (J-C-L) [reply src/main/java/com/aws/greengrass/componentmanager/ComponentServiceHelper.java:82]: Was wondering if we should add a new API. What's our process for deprecation APIs without breaking folks using them?

File: src/main/java/com/aws/greengrass/componentmanager/ComponentServiceHelper.java (hunk 1 above)
Comment (J-C-L) [reply src/main/java/com/aws/greengrass/componentmanager/ComponentServiceHelper.java:85]: That's cleaner. Updated.

File: src/main/java/com/aws/greengrass/componentmanager/ComponentServiceHelper.java (hunk 2 above)
Comment (J-C-L) [reply src/main/java/com/aws/greengrass/componentmanager/ComponentServiceHelper.java:82]: Spoke offline. Will add @Deprecated and create a new API.

File: src/test/java/com/aws/greengrass/componentmanager/ComponentServiceHelperTest.java [hunk 3]
```diff
@@ -60,7 +62,6 @@ class ComponentServiceHelperTest {
     @BeforeEach
     void beforeEach() {
         PlatformResolver platformResolver = new PlatformResolver(null);
-        lenient().when(clientFactory.getGreengrassV2DataClient()).thenReturn(client);
```
Comment (kucnavya) [inline src/test/java/com/aws/greengrass/componentmanager/ComponentServiceHelperTest.java:63]: Would it be useful to have some tests still cover this code path and test `getGreengrassV2DataClient` ?

File: src/main/java/com/aws/greengrass/deployment/DeploymentDocumentDownloader.java [hunk 4]
```diff
@@ -173,6 +175,7 @@ private GetDeploymentConfigurationResponse getDeploymentConfiguration(String dep
                     "Failed to contact Greengrass cloud or unable to parse response.", e);
         }
 
+
```
Comment (kucnavya) [inline src/main/java/com/aws/greengrass/deployment/DeploymentDocumentDownloader.java:178]: nit: remove new line

**Review by kucnavya — APPROVED** [review]

File: src/test/java/com/aws/greengrass/componentmanager/ComponentServiceHelperTest.java (hunk 3 above)
Comment (J-C-L) [reply src/test/java/com/aws/greengrass/componentmanager/ComponentServiceHelperTest.java:63]: Not sure which code paths you mean. No coverage was removed in this PR. I did notice that nothing in GreengrassServiceClientFactory is tested. So yeah, we should probably add those tests as a follow up.

Comment (kucnavya) [reply src/test/java/com/aws/greengrass/componentmanager/ComponentServiceHelperTest.java:63]: I meant that `getGreengrassV2DataClient()` is now replaced with `fetchGreengrassV2DataClient()` and tests are modified to use the new method. We could add some more tests that cover calls to `getGreengrassV2DataClient()`. But that can be a follow up.

**Pull request merged**